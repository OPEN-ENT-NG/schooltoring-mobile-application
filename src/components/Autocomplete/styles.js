import { StyleSheet } from "react-native";

import { COLORS } from "../../styles/common";

export default StyleSheet.create({
  input: {
    backgroundColor: COLORS.WHITE
  },
  list: {
    backgroundColor: COLORS.WHITE,
    zIndex: 10,
    position: "relative"
  },
  item: {
    padding: 20
  }
});
