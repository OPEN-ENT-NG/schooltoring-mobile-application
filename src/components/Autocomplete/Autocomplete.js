import React, { Component } from "react";
import {
  View,
  TextInput,
  ScrollView,
  Keyboard,
  Platform,
  Dimensions
} from "react-native";

import PropTypes from "prop-types";
import Touchable from "../Touchable/Touchable";

import styles from "./styles";
import { COLORS } from "../../styles/common";

export default class Autocomplete extends Component {
  constructor(props) {
    super(props);
    this.state = {
      filter: "",
      output: [],
      listHeight: Dimensions.get("window").height * 0.6
    };
    this.updateList = this.updateList.bind(this);
    this.renderItem = this.renderItem.bind(this);
    this.delta = 200;
  }

  componentDidMount() {
    if (Platform.OS === "ios") {
      this.keyboardWillShowEvent = Keyboard.addListener(
        "keyboardWillShow",
        this._updateKeyboardSpace
      );
      this.keyboardWillHideEvent = Keyboard.addListener(
        "keyboardWillHide",
        this._resetKeyboardSpace
      );
    } else if (Platform.OS === "android") {
      this.keyboardWillShowEvent = Keyboard.addListener(
        "keyboardDidShow",
        this._updateKeyboardSpace
      );
      this.keyboardWillHideEvent = Keyboard.addListener(
        "keyboardDidHide",
        this._resetKeyboardSpace
      );
    }
  }

  componentWillUnmount() {
    this.keyboardWillShowEvent && this.keyboardWillShowEvent.remove();
    this.keyboardWillHideEvent && this.keyboardWillHideEvent.remove();
  }

  updateList(filter) {
    let output = filter
      ? this.props.data.filter(item => this.props.filterItem(item, filter))
      : [];
    this.setState({ filter, output });
  }

  _updateKeyboardSpace = e => {
    this.setState({
      listHeight: this.state.listHeight - this.delta
    });
  };

  _resetKeyboardSpace = () => {
    this.setState({ listHeight: this.state.listHeight + this.delta });
  };

  renderItem(item) {
    return (
      <Touchable
        style={{ backgroundColor: COLORS.SECONDARY }}
        key={this.props.getItemKey(item)}
        onPress={() => {
          this.props.onItemPress(item);
          this.setState({ filter: "", output: [] });
        }}>
        <View style={styles.item}>{this.props.renderItem(item)}</View>
      </Touchable>
    );
  }

  render() {
    return (
      <View
        style={[this.props.style, Platform.OS == "ios" ? { zIndex: 100 } : {}]}>
        <TextInput
          placeholder={this.props.placeholder}
          onChangeText={filter => this.updateList(filter)}
          value={this.state.filter}
          style={styles.input}
          underlineColorAndroid={this.props.underlineColor}
        />
        {this.state.output.length > 0 && (
          <View
            style={[
              styles.scrollviewContainer,
              { height: this.state.listHeight }
            ]}>
            <ScrollView
              style={styles.scrollview}
              keyboardShouldPersistTaps="handled">
              {this.state.output.map(item => this.renderItem(item))}
            </ScrollView>
          </View>
        )}
      </View>
    );
  }
}

Autocomplete.propTypes = {
  placeholder: PropTypes.string,
  data: PropTypes.array.isRequired,
  filterItem: PropTypes.func.isRequired,
  getItemKey: PropTypes.func.isRequired,
  onItemPress: PropTypes.func.isRequired,
  renderItem: PropTypes.func.isRequired
};
