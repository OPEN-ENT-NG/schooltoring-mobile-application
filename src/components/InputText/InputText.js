import React, { Component } from "react";
import PropTypes from "prop-types";
import { TextInput, Platform } from "react-native";

import styles from "./styles";

const InputText = props => {
  return (
    <TextInput
      secureTextEntry={props.secureTextEntry || false}
      style={styles.input}
      placeholder={props.placeholder}
      value={props.value}
      onChangeText={val => props.onChangeText(val)}
    />
  );
};

InputText.propTypes = {
  secureTextEntry: PropTypes.bool,
  placeholder: PropTypes.string,
  value: PropTypes.string.isRequired,
  onChangeText: PropTypes.func.isRequired
};

export default InputText;
