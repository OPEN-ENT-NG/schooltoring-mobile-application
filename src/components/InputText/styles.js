import { StyleSheet, Platform } from "react-native";

import { COLORS } from "../../styles/common";

const styles = StyleSheet.create({
  input: {
    height: 50,
    ...Platform.select({
      ios: {
        borderWidth: 1,
        borderColor: COLORS.GREY,
        marginBottom: 15,
        padding: 10,
        borderRadius: 3
      }
    })
  }
});

export default styles;
