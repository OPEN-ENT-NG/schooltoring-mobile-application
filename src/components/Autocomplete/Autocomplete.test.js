import React from "react";
import { TextInput, ScrollView } from "react-native";
import { shallow } from "enzyme";

import Autocomplete from "./Autocomplete";
import InputText from "../InputText/InputText";

import state from "../../jest/state.json";

const placeholder = "placeholder text";

describe("Render", () => {
  const rendered = shallow(
    <Autocomplete
      placeholder={placeholder}
      data={state.subjects}
      filterItem={jest.fn()}
      getItemKey={jest.fn()}
      onItemPress={jest.fn()}
      renderItem={jest.fn()}
    />
  );

  test("Autocomplete should render without crashing", () => {
    expect(rendered).toMatchSnapshot();
  });

  test(`Placeholder text should equal '${placeholder}'`, () => {
    const input = rendered.find(InputText);
    expect(input.prop("placeholder")).toEqual(placeholder);
  });

  test("An error should be logged if no required props are given", () => {
    const PropTypes = jest.spyOn(global.console, "error");
    const component = shallow(<Autocomplete />);
    expect(component).toMatchSnapshot();
    expect(PropTypes).toHaveBeenCalled();
  });
});

describe("Interactions", () => {
  const filterItem = jest.fn(() => true),
    getItemKey = jest.fn(),
    onItemPress = jest.fn(),
    renderItem = jest.fn(),
    value = "ANG",
    data = [
      {
        subjectId: "1475566-1517929209848",
        subjectCode: "ZIP-220700",
        subjectLabel: "ESSAI DE SYSTEMES"
      },
      {
        subjectId: "1475710-1517929209848",
        subjectCode: "ZIP-341700",
        subjectLabel: "VENTE(TP,PRESENTATION,EXPR.,COMPORTEMENT"
      },
      {
        subjectId: "1475847-1517929209848",
        subjectCode: "ZIP-308600",
        subjectLabel: "CHIMIE-BIOCHIMIE-SCIENCES DU VIVANT"
      }
    ];
  const rendered = shallow(
    <Autocomplete
      placeholder={placeholder}
      data={data}
      filterItem={filterItem}
      getItemKey={getItemKey}
      onItemPress={onItemPress}
      renderItem={renderItem}
    />
  );
  let itemList;

  test("On change text should trigger filter item props", () => {
    const inputText = rendered.find(InputText);
    const input = inputText.dive().find(TextInput);
    input.value = value;
    inputText.props().onChangeText(value);
    expect(filterItem).toHaveBeenCalled();
  });

  test("Search event should display 3 items", () => {
    rendered.update();
    itemList = rendered.find(ScrollView).children();
    expect(itemList.length).toEqual(3);
  });

  test("Choose autocomplete item should trigger on item press props function", () => {
    const item = itemList.first();
    item.simulate("press");
    expect(onItemPress).toHaveBeenCalled();
  });

  test("Autocomplete list should empty after item selection", () => {
    rendered.update();
    itemList = rendered.find(ScrollView).children();
    expect(itemList.length).toEqual(0);
  });
});
