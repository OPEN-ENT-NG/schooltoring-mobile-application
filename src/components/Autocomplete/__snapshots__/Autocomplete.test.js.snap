// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render An error should be logged if no required props are given 1`] = `
<Component
  style={
    Array [
      undefined,
      undefined,
    ]
  }
>
  <InputText
    onChangeText={[Function]}
    style={
      Object {
        "backgroundColor": "#ffffff",
      }
    }
    value=""
  />
  <ScrollView
    keyboardShouldPersistTaps="handled"
    style={
      Array [
        Object {
          "backgroundColor": "#ffffff",
          "position": "relative",
          "zIndex": 10,
        },
        Object {
          "maxHeight": 933.8,
        },
      ]
    }
  />
</Component>
`;

exports[`Render Autocomplete should render without crashing 1`] = `
<Component
  style={
    Array [
      undefined,
      undefined,
    ]
  }
>
  <InputText
    onChangeText={[Function]}
    placeholder="placeholder text"
    style={
      Object {
        "backgroundColor": "#ffffff",
      }
    }
    value=""
  />
  <ScrollView
    keyboardShouldPersistTaps="handled"
    style={
      Array [
        Object {
          "backgroundColor": "#ffffff",
          "position": "relative",
          "zIndex": 10,
        },
        Object {
          "maxHeight": 933.8,
        },
      ]
    }
  />
</Component>
`;
